### PR TITLE
Fix caching issue during sync

### LINKED ISSUE #384: Certain movies "watched status" are being added multiple times to Plex history
### The problem

Certain movies "watched status" are being added multiple times, here's an example:

```
amaral@HYPERCUBE:~/.local/scripts/PlexTraktSync$ pipenv run python main.py sync --sync movies
Loading .env environment variables...
INFO: PlexTraktSync [5dfc1801: Merge pull request #382 from glensc/Pipfile.lock-regenerate @2021-07-14 11:58:37]
INFO: Syncing with Plex redacted@domain.com and Trakt <redacted>
Sync Movies: True
Sync Shows: False
INFO: Loaded Trakt lists in 3.8 seconds
Plex Server version: 1.23.5.4801-bc4dd58bb, updated at: 2021-07-15 12:30:49
Server has 2 libraries: ['Movies', 'TV Shows']
Processing Movies  [###---------------------------------]  22/229  00:00:23INFO: Marking as watched in Plex: <tmdb:497698:<Movie:5121:Black-Widow>>
Processing Movies  [###############---------------------]  100/229  00:00:14INFO: Marking as watched in Plex: <tmdb:512200:<Movie:3118:Jumanji:-The-Next-Le>>
Processing Movies  [###################################-]  226/229  00:00:00INFO: Marking as watched in Plex: <tmdb:637649:<Movie:3249:Wrath-of-Man>>
Processing Movies  [####################################]  229/229
INFO: Movies processed in 30.2 seconds
INFO: Updated plex watchlist in 0.0 seconds
INFO: Completed full sync in 34.1 seconds
amaral@HYPERCUBE:~/.local/scripts/PlexTraktSync$ pipenv run python main.py sync --sync movies
Loading .env environment variables...
INFO: PlexTraktSync [5dfc1801: Merge pull request #382 from glensc/Pipfile.lock-regenerate @2021-07-14 11:58:37]
INFO: Syncing with Plex <redacted@domain.com> and Trakt <redacted>
Sync Movies: True
Sync Shows: False
INFO: Loaded Trakt lists in 3.0 seconds
Plex Server version: 1.23.5.4801-bc4dd58bb, updated at: 2021-07-15 12:30:49
Server has 2 libraries: ['Movies', 'TV Shows']
Processing Movies  [###---------------------------------]  22/229  00:00:23INFO: Marking as watched in Plex: <tmdb:497698:<Movie:5121:Black-Widow>>
Processing Movies  [###############---------------------]  100/229  00:00:14INFO: Marking as watched in Plex: <tmdb:512200:<Movie:3118:Jumanji:-The-Next-Le>>
Processing Movies  [###################################-]  226/229  00:00:00INFO: Marking as watched in Plex: <tmdb:637649:<Movie:3249:Wrath-of-Man>>
Processing Movies  [####################################]  229/229
INFO: Movies processed in 30.4 seconds
INFO: Updated plex watchlist in 0.0 seconds
INFO: Completed full sync in 33.4 seconds
```

What do these specific movies have in common, you might ask? I simply went to Trakt.tv website, and added something to my history, like this:

image

So, **every time I run** `pipenv run python main.py sync --sync movies` (or just `pipenv run python main.py`), the movies that were added to my Trakt.tv history, are going to be added multiple times to my Plex history.

### Expected behavior

Movies/episodes should only be added **once** to the Plex watched history when syncing the history from the Trakt.tv website.

### Steps to reproduce the behavior

Not sure, I've only followed the instructions to install PlexTraktSync and configured my setup like this:

```json
{
  "log_debug_messages": false,
  "logging": {
    "append": false
  },
  "sync": {
    "liked_lists": false,
    "watchlist": false,
    "watched_status": true,
    "collection": false,
    "ratings": false
  },
  "xbmc-providers": {
    "movies": "imdb",
    "shows": "tvdb"
  },
  "excluded-libraries": [
    "Samples"
  ]
}
```

### Workarounds

Avoid running the script and mark "as played" manually on Plex.

### Environment

- Version: `main`
- Python version: `3.8.8`
- Platform: `synology_apollolake_918+`

### PR DESCRIPTION
**Pull request by mcdallas**
Fixes #384

### CODE REVIEW

File: plex_trakt_sync/media.py [hunk 1]
```diff
@@ -13,13 +14,17 @@ class Media:
...
 
-    def __init__(self, plex, trakt, plex_api: PlexApi = None, trakt_api: TraktApi = None):
+    def __init__(self, plex: PlexLibraryItem, trakt, plex_api: PlexApi = None, trakt_api: TraktApi = None):
         self.plex_api = plex_api
         self.trakt_api = trakt_api
         self.plex = plex
         self.trakt = trakt
         self.show = None
 
+    @nocache
```
Comment (glensc) [inline plex_trakt_sync/media.py:24]: remove @nocache here, it should be on actual method making the call, i.e fetch_item

**Comment by glensc** [comment]
Have you compared the performance of this? as this looks, it will make a single API call for each item in the library, so this will be very slow.

when comparing compare with pre-0.13.0 version as well, and movie and tv library types.

if that is so, then perhaps close this PR and do another PR with `@nocache` added to the call that fetches library contents (large XML with paging, but with all movie details).

**Comment by mcdallas** [comment]
> if that is so, then perhaps close this PR and do another PR with @nocache added to the call that fetches library contents (large XML with paging, but with all movie details).

could you point out where this happens in the code?

**Comment by glensc** [comment]
@mcdallas did you compare the performance? share your details!

looked up from the code, it's `all()` that retrieves the XML in chunks, but it already has the `@nocache` decorator:
- https://github.com/Taxel/PlexTraktSync/blob/0.13.5/plex_trakt_sync/plex_api.py#L361-L363

I guess you have to enable debug logs and then inspect log file...

**Comment by mcdallas** [comment]
The performance goes from 9.5s to 14.5s to process 350 movies

**Comment by glensc** [comment]
those stats are with cold cache both? please share stats for both cold and hot cache with before and after. and detail which agent you tested with.

also, with the rebase, you've lost the `@nocache` directive.

**Comment by mcdallas** [comment]
> also, with the rebase, you've lost the `@nocache` directive.

I've reused the existing `reload_item` method and removed nocache from fetch_item

**Comment by glensc** [comment]
`reload_item` does not have `@nocache` currently, so it will be cached by `requests_cache`.

**Comment by mcdallas** [comment]
Well the fix works regardless, I don't get duplicates

**Comment by mcdallas** [comment]
If I delete trakt_cache.sqlite before every run:
with this PR `INFO: Movies processed in 1 min 20.0 seconds`
without this PR `INFO: Movies processed in 1 min 3.7 seconds`

without deleting trakt_chache.sqlite:
with this PR `INFO: Movies processed in 14.8 seconds`
without this PR `INFO: Movies processed in 9.5 seconds`

**Comment by glensc** [comment]
post debug log of the reload_item call. is it perhaps POST request and therefore not cached? 

I'm very sure the reload_item call is cached:

run `./plex_trakt_sync.sh cache --limit=0 | grep YOUR_ID` to see if such request is cached.

are you sure you're testing the right thing? try running first when the item is not played, with your code from this branch. set item played, and it will use cached request for reload_item.

**Comment by mcdallas** [comment]
Whether reload_item or fetch_item is being cached is irrelevant. The issue is that section.items() return stale data.

I am running this script
```python
from plex_trakt_sync.factory import factory

MOVIE_ID = '6077'

api = factory.plex_api()

sections = api.movie_sections()

for movie in sections[0].items():
    if movie.id == MOVIE_ID:
        print(f"Movie Title: {movie.item.title}")
        print(f"IsWatched: {movie.item.isWatched}")
        reloaded = api.reload_item(movie)
        print(f"Reloaded isWatched: {reloaded.item.isWatched}")
        break
```

```
> python -W ignore cache_test.py 
Movie Title: 13 Tzameti
IsWatched: False
Reloaded isWatched: True
```

The correct status of this movie is watched (it shows up as watched on plex) but the movie object returned from sections.items() says it's not watched

**Comment by glensc** [comment]
I don't understand why fetch_item being in HTTP cache is irrelevant. first time it's invoked (for every move) when the movie is not played -> cached as notSeen, once it's played, and next run calls fetch_item, entry taken from cache -> result is notSeen.

I'll look into your explanation and test script tomorrow (maybe).

**Comment by glensc** [comment]
update your test case so that the first-time reload_item is called the movie is not played (remove played state from plex for that).

**Comment by glensc** [comment]
and anyway, as reload_item is called for EVERY item in the library, this Mr should not be merged as it degrades performance. I said this earlier too

**Comment by glensc** [comment]
so, you tested wrong, or is something else is happening?

also, be transparent and include reasoning why you closed, so others can understand that you did not click the wrong button.

**Comment by glensc** [comment]
regarding `@nocache` for `reload_item`. made a test

1. ran inspect 8057
2. it prints watched on plex: true
3. removed watched status on plex
4. applied the diff below
5. ran inspect 8057, in prints watched on plex: true

so, `@nocache` for `reload_item` is needed. but as you closed this PR, I'm gonna submit this as separate PR.

```diff
➔ g diff
diff --git a/plex_trakt_sync/commands/inspect.py b/plex_trakt_sync/commands/inspect.py
index c68b8e2..8de44b4 100644
--- a/plex_trakt_sync/commands/inspect.py
+++ b/plex_trakt_sync/commands/inspect.py
@@ -21,6 +21,7 @@ def inspect(input):
         input = int(input)

     pm = plex.fetch_item(input)
+    pm = plex.reload_item(pm)
     print(f"Inspecting: {pm}")

     url = plex.media_url(pm)
```

**Comment by glensc** [comment]
also, you did not answer what agent are you using.

seems with New Movie Agent the `@nocache` on `all()` has no effect because plexapi internally does some "reloads":

```
2021-08-01 10:02:36,680 DEBUG[plexapi]:Reloading Movie 'Movie Name' for attr 'guids'
```

and that reload is happening outside `all()`, i.e in `guid` method which does not have `@nocache`.

**Comment by glensc** [comment]
Added helper to be able to see cached request content:
- https://github.com/Taxel/PlexTraktSync/pull/405/files

for me 8057 has only guid attribute (no guids) as I'm using legacy agent:

```xml

<Video ratingKey="8057" key="/library/metadata/8057" guid="com.plexapp.agents.imdb://ttXXXXXX?lang=en"
```

**Comment by mcdallas** [comment]
> be transparent and include reasoning why you closed, so others can understand that you did not click the wrong button.

You said you won’t merge the PR so I closed it

> you did not answer what agent are you using.

I am using the new agent

**Comment by glensc** [comment]
Can you try this instead:
- https://github.com/Taxel/PlexTraktSync/pull/410

**Pull request closed without merging**